### PR TITLE
Prevent loading HD5Stores into memory

### PR DESCRIPTION
**Pull request by wassname**
The command np.asarray loads the whole thing into memory causing an out of memory error when using bcolz, HD5Store or anything similar. 

This changes it to only convert type if needed.

### CODE REVIEW

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -771,7 +771,9 @@ def __init__(self, x, y, image_data_generator,
 
         if data_format is None:
             data_format = K.image_data_format()
-        self.x = np.asarray(x, dtype=K.floatx())
+        if not getattr(x, 'dtype', None) != K.floatx():
```
Comment (fchollet) [inline keras/preprocessing/image.py]: `if not {expression} != {expression}:` is quite a strange structure. Did you mean:

```python
dtype = getattr(x, 'dtype', None)
if dtype != K.floatx():
   x = np.asarray(x, dtype=K.floatx())
```

Comment (wassname) [reply keras/preprocessing/image.py]: Yeah that's looks nicer, I'll change it

**Comment by wassname** [comment]
Ok will do

**Comment by lucasdavid** [comment]
According to [np.asarray docs](https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html):
> If dtype is set, array is copied only if dtype does not match

They even have a use case showing `asarray` will return the same reference when dtype is the same. Do you have an example which isn't handled correctly by `asarray`?

**Comment by lucasdavid** [comment]
Also: notice lines 770 and 793 have `np.asarray` calls which were not protected by your dtype checking.

**Comment by wassname** [comment]
Good points. I did have a case where it was trying to convert an on-disc bcolz array if I didn't do this, but perhaps it's only with that. This seems to be an edge case no one else has experienced (and I no longer have a problem with) so I reckon we close this issue. Thanks for digging up that material on dtype conversion, TIL.

**Comment by lucasdavid** [comment]
I see. Maybe the problem you faced was that the default numpy `dtype` is `float64` (at least in my computers), while `K.floatx()` default is `float32`? In this case, data'd be copied even after your changes (see execution bellow), and maybe the way to go is to try to have the data in `float32` since the start.

```python
import numpy as np
from keras.preprocessing.image import ImageDataGenerator

for dtype in ('float32', 'float64'):
    x = np.random.rand(32, 8, 8, 3).astype(dtype)
    y = np.random.randint(10, size=(32, 1))

    i = ImageDataGenerator().flow(x, y)

    print('The iterator', 'did not copied' if i.x is x else 'did copied',
          '`x`, dtypes: %s~>%s' % (x.dtype, i.x.dtype))
```
```shell
Using TensorFlow backend.
The iterator did not copied `x`, dtypes: float32~>float32
The iterator did copied `x`, dtypes: float64~>float32
```

**Pull request closed without merging**